langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java

```diff
@@ -98,7 +95,72 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
             return removeNulls(STRING, enums((Object[]) type.getEnumConstants()), description);
         }
 
-        return removeNulls(OBJECT, description); // TODO provide internals
+        return removeNulls(OBJECT, schema(type), description);
+    }
+
+    private static JsonSchemaProperty schema(Class<?> structured){
+        //TODO: Externalize the depth
+        return schema(structured, new HashSet<>(), 3);
+    }
+
+    private static JsonSchemaProperty schema(Class<?> structured, Set<Class<?>> visited, int level) {
+        if (level < 0 || visited.contains(structured)) {
+            return null;
+        }
+
+        visited.add(structured);
+        Map<String,Object> properties = new HashMap<>();
+        for (Field field : structured.getDeclaredFields()) {
+            String name = field.getName();
+            if (name.equals("__$hits$__") || name.equals("this$0") || java.lang.reflect.Modifier.isStatic(field.getModifiers())) {
```

Comment:
what is `this$0`?
could you pelase add a comment?

Reply:
done, inner classes would be represent as a field `this$0` we need to skip that.

Reply:
Ah right, thanks  for explanation!